dart/constraint/ConstraintSolver.hpp

```diff
@@ -89,6 +89,12 @@ class ConstraintSolver
   /// Add a constraint
   void addConstraint(const ConstraintBasePtr& constraint);
 
+  /// Returns all the constraints added to this ConstraintSolver.
+  const std::vector<constraint::ConstraintBasePtr>& getConstraints();
```

Comment:
Maybe I'm not thinking about this the right way, but shouldn't the non-const overload return a non-const reference to the vector while the const overload returns a const reference to the vector (instead of a copy of the vector)? Creating a copy should only be done if:

1. The underlying implementation does not contain a vector object that can be directly referenced
2. The API does not want to expose the implementation detail that the constraint pointers are held in a vector, in case we want to change this implementation in the future

Condition (1) doesn't apply here, and if we want to satisfy condition (2) then both the const and non-const overloads should both return a copy.

Reply:
Agreed. I think we should go with (2) since letting the caller to modify the list of constraints would be risky. Instead, we should enforce them to use the constraint modifier functions (`add/removeConstraint`).

The motivation this non-const overload returns const reference was to provide a way to iteratate the constraints or get the number of them. For this, it'd be good to provide `getNumConstraint()` and `getConstraint(index)`.